community/dbms/src/main/java/org/neo4j/commandline/dbms/DumpCommand.java

```diff
@@ -0,0 +1,203 @@
+        {
+            super( "dump" );
+        }
+
+        @Override
+        public Optional<String> arguments()
+        {
+            return Optional.of( "--database=<database> --to=<destination-path>" );
+        }
+
+        @Override
+        public String description()
+        {
+            return "Dump a database into a single-file archive. The archive can be used by the load command. " +
+                    "<destination-path> can be a file or directory (in which case a file called <database>.dump will " +
+                    "be created). It is not possible to dump a database that is mounted in a running Neo4j server.";
+        }
+
+        @Override
+        public AdminCommand create( Path homeDir, Path configDir, OutsideWorld outsideWorld )
+        {
+            return new DumpCommand( homeDir, configDir, new Dumper() );
+        }
+    }
+
+    private final Path homeDir;
+    private final Path configDir;
+    private final Dumper dumper;
+
+    public DumpCommand( Path homeDir, Path configDir, Dumper dumper )
+    {
+        this.homeDir = homeDir;
+        this.configDir = configDir;
+        this.dumper = dumper;
+    }
+
+    @Override
+    public void execute( String[] args ) throws IncorrectUsage, CommandFailed
+    {
+        String database = parse( args, "database", identity() );
+        Path archive = calculateArchive( database, parse( args, "to", Paths::get ) );
+        Path databaseDirectory = toDatabaseDirectory( database );
+
+        try ( Closeable ignored = withLock( databaseDirectory ) )
+        {
+            dump( database, databaseDirectory, archive );
+        }
+        catch ( StoreLockException e )
+        {
+            throw new CommandFailed( "the database is in use -- stop Neo4j and try again", e );
+        }
+        catch ( IOException e )
+        {
+            wrapIOException( e );
+        }
+    }
+
+    private <T> T parse( String[] args, String argument, Function<String, T> converter ) throws IncorrectUsage
+    {
+        try
+        {
+            return Args.parse( args ).interpretOption( argument, mandatory(), converter );
+        }
+        catch ( IllegalArgumentException e )
+        {
+            throw new IncorrectUsage( e.getMessage() );
+        }
+    }
+
+    private Path toDatabaseDirectory( String databaseName )
+    {
+        //noinspection unchecked
+        return new ConfigLoader( asList( DatabaseManagementSystemSettings.class, GraphDatabaseSettings.class ) )
+                .loadConfig(
+                        Optional.of( homeDir.toFile() ),
+                        Optional.of( configDir.resolve( "neo4j.conf" ).toFile() ) )
+                .with( stringMap( DatabaseManagementSystemSettings.active_database.name(), databaseName ) )
+                .get( database_path ).toPath();
+    }
+
+    private Path calculateArchive( String database, Path to )
+    {
+        Path archive;
+        if ( Files.exists( to ) && Files.isDirectory( to ) )
```

Comment:
I think `#isDirectory()` checks for existence and `#exists()` call is not needed.